### PR TITLE
Issue #7809: Parallel Partition Sorting 

### LINKED ISSUE #7809: window performance
### What happens?

window functions seem to be slow, there was a previous issue here https://github.com/duckdb/duckdb/issues/1367
i thought i should re-raise because this may help fix the issue.
pandas takes 1 sec, duckdb takes 30, maybe its worth checking the logic in pandas and see if you can do the same? i think its here https://github.com/pandas-dev/pandas/blob/main/pandas/_libs/window/aggregations.pyx

### To Reproduce

file:
https://d37ci6vzurychx.cloudfront.net/trip-data/fhvhv_tripdata_2021-04.parquet

duckdb:
```
select
    sum(driver_pay) over _win,
    dropoff_datetime
from read_parquet("~/dev/data/taxi3/fhvhv_tripdata_2021-04.parquet")
WINDOW _win as (
        order by dropoff_datetime asc
        range between
        interval 3 days preceding and
        interval 0 days following
    )
)
```
pandas:
```
    df = pd.read_parquet(
        file1,
        columns=["driver_pay", "dropoff_datetime"],
    )
    df.index = pd.DatetimeIndex(df["dropoff_datetime"])
    df = df.sort_index()
    df = df.drop("dropoff_datetime", axis=1)
    df = df.rolling("3d").sum()
```

### OS:

linux

### DuckDB Version:

0.8

### DuckDB Client:

latest

### Full Name:

harry

### Affiliation:

harry

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

### PR DESCRIPTION
**Pull request by hawkfish**
* Refactor the partitioning code to support parallel scans and sorts.
* Implement multiple sort tasks.
* Fix jump reset code in WindowExecutor to handle reverse-order blocks.

### CODE REVIEW

**Comment by krlmlr** [comment]
Please skip the failing R check for now, I'll handle it downstream.

If possible, apply #8608 first.

**Review by krlmlr — APPROVED** [review]
Thanks. I've made a note to update duckplyr accordingly.

**Comment by hannes** [comment]
I think this is something for @lnkuiper to review once he's back

**Comment by hawkfish** [comment]
> I think this is something for @lnkuiper to review once he's back

When would that be? I was hoping to get it into 0.9.0.

File: test/sql/window/test_window_sort.test_slow [hunk 1]
```diff
@@ -69,7 +69,9 @@ select * from rownum_window order by l_extendedprice
 ----
 
 query IIIII nosort q3
-select l_extendedprice, list_value(l_shipdate), list_value(l_commitdate), list_value(l_receiptdate), row_number() over (order by l_extendedprice) from lineitem_d;
+select l_extendedprice, list_value(l_shipdate), list_value(l_commitdate), list_value(l_receiptdate), row_number() over (order by l_extendedprice) 
+from lineitem_d
+order by l_extendedprice;
```
Comment (lnkuiper) [inline test/sql/window/test_window_sort.test_slow:74]: I am a bit concerned with this, as it seems that the parallel scan of the window operator does not yield deterministic results. If someone executes this query, they would expect the output to be ordered by `l_extendedprice` as per the window specification and not have to add another `order by`.

We can have a deterministic output order for parallel scans using _batch indices_, which the window operator also needs. I have implemented this in the `PhysicalOrder` operator. You can check there for an example. There, the `batch_index` depends on the `block_idx` of the `PayloadScanner`.

The general idea of batch indices is that you assign an index to a range of values that are assigned to a thread. For example:

| thread | chunks | batch index |
|:--|:--|:--|
| 0 | 0-10 | 0 |
| 1 | 10-20 | 1 |
| 2 | 20-30 | 2 |
| 4 | 30-40 | 3 |

Thread 0 can keep outputting chunks using the same batch index of 0, and they will be appended in order to a `BatchedDataCollection` further in the pipeline if order matters there.

The number of chunks per batch is up to you. See what works best.

**Review by lnkuiper — CHANGES_REQUESTED** [review]
Hi @hawkfish, I'm back 😄. This mostly looks good, I have one remark before this can be merged:

File: test/sql/window/test_window_sort.test_slow (hunk 1 above)
Comment (hawkfish) [reply test/sql/window/test_window_sort.test_slow:74]: Thanks - this is a good point. 

The big question for me is how to handle work stealing. Would it make sense to assign a batch index to each work block (returned by `GetScanner`) and set the order up to be (partition, block)?

Comment (hawkfish) [reply test/sql/window/test_window_sort.test_slow:74]: I tried this and the problem is that work stealing results in threads potentially working on "earlier" batch indexes that they steal from skewed partitions.

Comment (hawkfish) [reply test/sql/window/test_window_sort.test_slow]: Talked to @Mytherin and we agreed that it is only reasonable to expected ordered output when there is a single partition. Got that working, but anything more complex will need to wait until the accelerator (e.g., segment tree) construction has been parallelised.

**Comment by Mytherin** [comment]
Thanks! LGTM

**Pull request merged**